storage: panic in raftGroup.Step while running under stressrace

Running at head (150b52daca1127372c6e925a164b9370586a570b), I was stressing `TestRemovePlaceholderRace` on an azworker to try to repro #12574 while looking into #14193. After running 5606 runs for over an hour, I hit a different (but possibly related) panic:

```
E170317 15:36:04.978177 166 vendor/github.com/coreos/etcd/raft/log.go:191  [s2,r1/24:{-},@c4208f9880] tocommit(123) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost?
panic: tocommit(123) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost? [recovered]
	panic: tocommit(123) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost?

goroutine 166 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc42001ebe0)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:198 +0xd8
panic(0x2135780, 0xc420543df0)
	/home/alex/go1.8/src/runtime/panic.go:489 +0x2f0
github.com/cockroachdb/cockroach/pkg/storage.(*raftLogger).Panicf(0xc420543b50, 0x23bf79b, 0x5d, 0xc420c8c460, 0x2, 0x2)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/storage/raft.go:111 +0x11d
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*raftLog).commitTo(0xc420c95500, 0x7b)
	/home/alex/go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/log.go:191 +0x1fe
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*raft).handleHeartbeat(0xc4201524b0, 0x8, 0x18, 0x1, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/alex/go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/raft.go:1100 +0x7a
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.stepFollower(0xc4201524b0, 0x8, 0x18, 0x1, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/alex/go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/raft.go:1046 +0x412
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*raft).Step(0xc4201524b0, 0x8, 0x18, 0x1, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/alex/go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/raft.go:778 +0x10e5
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*RawNode).Step(0xc42056bf80, 0x8, 0x18, 0x1, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/alex/go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/rawnode.go:195 +0x11e
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRaftRequest.func4(0xc42056bf80, 0x0, 0x0, 0x0)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3144 +0xf1
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).withRaftGroupLocked(0xc4208f9880, 0x23d3300, 0xc42074d6f0, 0xc4208f9988, 0xc)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:539 +0x1c2
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).withRaftGroup(0xc4208f9880, 0xc42074d6f0, 0x0, 0x0)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:556 +0xa7
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRaftRequest(0xc420219880, 0x2b2220281800, 0xc42039d140, 0xc4204c8900, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3145 +0x11aa
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRequestQueue(0xc420219880, 0x1)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3385 +0x1f0
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc4202102d0, 0xc42001ebe0)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:225 +0x30c
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2()
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:165 +0x4b
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc42001ebe0, 0xc4201b63e0)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:209 +0x8b
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:210 +0x74
```

All I ran to kick off the testing was:

```
make stressrace PKG=./storage/ TESTS=TestRemovePlaceholderRace STRESSFLAGS='-stderr -maxfails 1 -p 32'
```

Full logs from the run: 
[raft_panic.txt](https://github.com/cockroachdb/cockroach/files/851301/raft_panic.txt)

@bdarnell @petermattis

--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -703,8 +703,7 @@ func (r *Replica) destroyDataRaftMuLocked(
 	}
 	clearTime := timeutil.Now()
 
-	// Save a tombstone. The range cannot be re-replicated onto this node
-	// without having a replica ID of at least consistentDesc.NextReplicaID.
+	// Save a tombstone to ensure that replica IDs never get reused.
 	if err := r.setTombstoneKey(ctx, batch, &consistentDesc); err != nil {
 		return err
 	}
@@ -735,18 +734,43 @@ func (r *Replica) cancelPendingCommandsLocked() {
 	r.mu.proposals = map[storagebase.CmdIDKey]*ProposalData{}
 }
 
+// setTombstoneKey writes a tombstone to disk to ensure that replica IDs never
+// get reused. It determines what the minimum next replica ID can be using
+// the provided RangeDescriptor and the Replica's own ID.
+//
+// We have to be careful to set the right key, since a replica can be using an
+// ID that it hasn't yet received a RangeDescriptor for if it receives raft
+// requests for that replica ID (as seen in #14231).
 func (r *Replica) setTombstoneKey(
 	ctx context.Context, eng engine.ReadWriter, desc *roachpb.RangeDescriptor,
 ) error {
 	r.mu.Lock()
-	r.mu.minReplicaID = desc.NextReplicaID
+	nextReplicaID := r.nextReplicaIDLocked(desc)
+	r.mu.minReplicaID = nextReplicaID
 	r.mu.Unlock()
 	tombstoneKey := keys.RaftTombstoneKey(desc.RangeID)
 	tombstone := &roachpb.RaftTombstone{
-		NextReplicaID: desc.NextReplicaID,
+		NextReplicaID: nextReplicaID,
 	}
 	return engine.MVCCPutProto(ctx, eng, nil, tombstoneKey,
 		hlc.Timestamp{}, nil, tombstone)
+}
+
+// nextReplicaIDLocked returns the minimum ID that a new replica can be created
+// with for this replica's range. We have to be very careful to ensure that
+// replica IDs never get re-used because that can cause panics.
+//
+// The externalDesc parameter is an optional way to provide an additional
+// descriptor for the range that was looked up outside the replica code.
+func (r *Replica) nextReplicaIDLocked(externalDesc *roachpb.RangeDescriptor) roachpb.ReplicaID {
+	result := r.mu.state.Desc.NextReplicaID
+	if externalDesc != nil && externalDesc.NextReplicaID > result {
+		result = externalDesc.NextReplicaID
+	}
+	if r.mu.minReplicaID > result {
+		result = r.mu.minReplicaID
+	}
+	return result
 }
 
 func (r *Replica) setReplicaID(replicaID roachpb.ReplicaID) error {
@@ -778,7 +802,10 @@ func (r *Replica) setReplicaIDLocked(replicaID roachpb.ReplicaID) error {
 	// }
 
 	previousReplicaID := r.mu.replicaID
-	r.mu.replicaID = replicaID
+	r.mu.replicaID = replicaID // TODO(DONOTMERGE): Should we update the log string here? Or keep the log string consistent with the descriptor?
+	if replicaID >= r.mu.minReplicaID {
+		r.mu.minReplicaID = replicaID + 1
+	}
 	// Reset the raft group to force its recreation on next usage.
 	r.mu.internalRaftGroup = nil
 

--- a/pkg/storage/replica_test.go
+++ b/pkg/storage/replica_test.go
@@ -6208,16 +6208,17 @@ func TestSetReplicaID(t *testing.T) {
 	repl := tc.repl
 
 	testCases := []struct {
-		replicaID    roachpb.ReplicaID
-		minReplicaID roachpb.ReplicaID
-		newReplicaID roachpb.ReplicaID
-		expected     string
+		replicaID            roachpb.ReplicaID
+		minReplicaID         roachpb.ReplicaID
+		newReplicaID         roachpb.ReplicaID
+		expectedMinReplicaID roachpb.ReplicaID
+		expectedErr          string
 	}{
-		{0, 0, 1, ""},
-		{0, 1, 1, ""},
-		{0, 2, 1, "raft group deleted"},
-		{1, 2, 1, ""}, // not an error; replicaID == newReplicaID is checked first
-		{2, 0, 1, "replicaID cannot move backwards"},
+		{0, 0, 1, 2, ""},
+		{0, 1, 1, 2, ""},
+		{0, 2, 1, 2, "raft group deleted"},
+		{1, 2, 1, 2, ""}, // not an error; replicaID == newReplicaID is checked first
+		{2, 0, 1, 0, "replicaID cannot move backwards"},
 	}
 	for _, c := range testCases {
 		t.Run("", func(t *testing.T) {
@@ -6227,8 +6228,13 @@ func TestSetReplicaID(t *testing.T) {
 			repl.mu.Unlock()
 
 			err := repl.setReplicaID(c.newReplicaID)
-			if !testutils.IsError(err, c.expected) {
-				t.Fatalf("expected %q, but found %v", c.expected, err)
+			repl.mu.Lock()
+			if repl.mu.minReplicaID != c.expectedMinReplicaID {
+				t.Errorf("expected minReplicaID=%d, but found %d", c.expectedMinReplicaID, repl.mu.minReplicaID)
+			}
+			repl.mu.Unlock()
+			if !testutils.IsError(err, c.expectedErr) {
+				t.Fatalf("expected %q, but found %v", c.expectedErr, err)
 			}
 		})
 	}

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -3126,7 +3126,7 @@ func (s *Store) processRaftRequest(
 			// Raft has decided the snapshot shouldn't be applied we would be
 			// writing the tombstone key incorrectly.
 			r.mu.Lock()
-			r.mu.minReplicaID = r.mu.state.Desc.NextReplicaID
+			r.mu.minReplicaID = r.nextReplicaIDLocked(nil)
 			r.mu.Unlock()
 		}
 

--- a/pkg/storage/store_test.go
+++ b/pkg/storage/store_test.go
@@ -2450,6 +2450,84 @@ func TestStoreRemovePlaceholderOnRaftIgnored(t *testing.T) {
 	})
 }
 
+// Test that we set proper tombstones for removed replicas and use the
+// tombstone to reject attempts to create a replica with a lesser ID.
+func TestRemovedReplicaTombstone(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	const rangeID = 1
+	creatingReplica := roachpb.ReplicaDescriptor{
+		NodeID:    2,
+		StoreID:   2,
+		ReplicaID: 2,
+	}
+
+	// All test cases assume that the starting replica ID is 1. This assumption
+	// is enforced by a check within the test logic.
+	testCases := []struct {
+		setReplicaID      roachpb.ReplicaID // set the existing replica to this before removing it
+		descNextReplicaID roachpb.ReplicaID // the descriptor's NextReplicaID during replica removal
+		createReplicaID   roachpb.ReplicaID // try creating a replica at this ID
+		expectCreated     bool
+	}{
+		{1, 2, 2, true},
+		{1, 2, 1, false},
+		{1, 2, 1, false},
+		{1, 3, 1, false},
+		{1, 3, 2, false},
+		{1, 3, 3, true},
+		{1, 99, 98, false},
+		{1, 99, 99, true},
+		{2, 2, 2, false},
+		{2, 2, 3, true},
+		{2, 2, 99, true},
+		{98, 2, 98, false},
+		{98, 2, 99, true},
+	}
+	for _, c := range testCases {
+		t.Run("", func(t *testing.T) {
+			tc := testContext{}
+			stopper := stop.NewStopper()
+			defer stopper.Stop()
+			tc.Start(t, stopper)
+			s := tc.store
+
+			// Bump up the existing replica's ID as specified then remove it to get a
+			// tombstone written.
+			repl1, err := s.GetReplica(rangeID)
+			if err != nil {
+				t.Fatal(err)
+			}
+			repl1.mu.Lock()
+			if repl1.mu.replicaID != 1 {
+				repl1.mu.Unlock()
+				t.Fatalf("test precondition not met; expected ReplicaID=1, got %d", repl1.mu.replicaID)
+			}
+			repl1.mu.Unlock()
+			if err := repl1.setReplicaID(c.setReplicaID); err != nil {
+				t.Fatal(err)
+			}
+			desc := roachpb.RangeDescriptor{
+				RangeID:       rangeID,
+				NextReplicaID: c.descNextReplicaID,
+			}
+			if err := s.RemoveReplica(context.Background(), repl1, desc, true); err != nil {
+				t.Fatal(err)
+			}
+
+			_, created, err := s.getOrCreateReplica(rangeID, c.createReplicaID, &creatingReplica)
+			if created != c.expectCreated {
+				t.Errorf("expected s.getOrCreateReplica(%d, %d, %v).created=%v, got %v",
+					rangeID, c.createReplicaID, creatingReplica, c.expectCreated, created)
+			}
+			if !c.expectCreated && !testutils.IsError(err, "raft group deleted") {
+				t.Errorf("expected s.getOrCreateReplica(%d, %d, %v).err='raft group deleted', got %v",
+					rangeID, c.createReplicaID, creatingReplica, err)
+			}
+		})
+	}
+}
+
 func TestCanCampaignIdleReplica(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	tc := testContext{}

